crates/apub/src/http/inbox.rs

```diff
@@ -0,0 +1,171 @@
+use crate::{
+  activities::WithPublished,
+  activity_lists::SharedInboxActivities,
+  fetcher::user_or_community::UserOrCommunity,
+};
+use activitypub_federation::{
+  actix_web::inbox::{receive_activity, receive_activity_parts},
+  config::Data,
+};
+use actix_web::{http::header::HeaderMap, web::Bytes, HttpRequest, HttpResponse};
+use chrono::{DateTime, Local, TimeDelta, Utc};
+use http::{Method, Uri};
+use lemmy_api_common::context::LemmyContext;
+use lemmy_utils::error::LemmyResult;
+use once_cell::sync::Lazy;
+use std::{
+  cmp::Ordering,
+  collections::BinaryHeap,
+  sync::{Arc, RwLock},
+  thread::available_parallelism,
+  time::Duration,
+};
+use tokio::{spawn, task::JoinHandle, time::sleep};
+use tracing::info;
+
+/// Handle incoming activities.
+pub async fn shared_inbox(
+  request: HttpRequest,
+  bytes: Bytes,
+  data: Data<LemmyContext>,
+) -> LemmyResult<HttpResponse> {
+  match serde_json::from_slice::<WithPublished<()>>(&bytes)?.published {
+    Some(published) => {
+      // includes published timestamp, insert to queue to ensure that activities are processed
+      // in correct order even when delivered out of order.
+      let request_parts = (
+        request.headers().clone(),
+        request.method().clone(),
+        request.uri().clone(),
+      );
+      ACTIVITY_QUEUE.write().unwrap().push(InboxActivity {
+        request_parts,
+        bytes,
+        published,
+      });
+      Ok(HttpResponse::Ok().finish())
+    }
+    None => {
+      // no timestamp included, process immediately
+      receive_activity::<SharedInboxActivities, UserOrCommunity, LemmyContext>(
+        request, bytes, &data,
+      )
+      .await
+    }
+  }
+}
+
+/// Queue of incoming activities, ordered by oldest published first
+static ACTIVITY_QUEUE: Lazy<Arc<RwLock<BinaryHeap<InboxActivity>>>> =
+  Lazy::new(|| Arc::new(RwLock::new(BinaryHeap::new())));
```

Comment:
There's an issue here with missing backpressure: The activity sending system intentionally has backpressure as in when you can't process activities quickly enough your responses will get slower which makes the sender send slower.

Since this heap is unbounded and the HTTP server immediately responds with Ok, regardless of whether the event has been processed, this means that if you are 1M events behind with this change the receiving queue will start receiving events at full speed until all million activities are in RAM here, regardless of how fast you're actually processing them.

This will both cause OOM issues as well as losing all of those events which makes the queue unreliable. 

There needs to be backpressure of some form - for example, the heap insert should wait until there's less than a limited amount of items in the queue.

Reply:
Thats true. Then the solution would be to have a hashmap with one heap per instance, and if the heap already contains > 5 activities for this origin instance, sleep in http handler until it decreases and then insert.